File: editorial_ch/chap4-1/prob4-1-5.py
# 向量 (ax, ay) 與 (bx, by) 的外積大小
def cross(ax, ay, bx, by):
	return ax * by - ay * bx

# 輸入
X1, Y1 = map(int, input().split())
X2, Y2 = map(int, input().split())
X3, Y3 = map(int, input().split())
X4, Y4 = map(int, input().split())

# 計算 cross 的值
ans1 = cross(X2-X1, Y2-Y1, X3-X1, Y3-Y1);
ans2 = cross(X2-X1, Y2-Y1, X4-X1, Y4-Y1);
ans3 = cross(X4-X3, Y4-Y3, X1-X3, Y1-Y3);
ans4 = cross(X4-X3, Y4-Y3, X2-X3, Y2-Y3);

# 全部排在一直線上時（邊角案例）
if ans1 == 0 and ans2 == 0 and ans3 == 0 and ans4 == 0:
	# 將　A, B, C, D 視為數值（正確來說是 pair 型態）
	# 藉由適當的進行 swap ，可以假設 A<B, C<D
	# 如此，可以歸結成判斷區間是否重疊的問題（節末問題 2.5.6）
	A = (X1, Y1) # 點 A 的座標
	B = (X2, Y2) # 點 B 的座標
	C = (X3, Y3) # 點 C 的座標
	D = (X4, Y4) # 點 D 的座標
	if A > B:
		tmp = B
		B = A
		A = tmp
	if C > D:
		tmp = D
		D = C
		C = tmp
	if max(A, C) <= min(B, D):
		print("Yes")
	else:
		print("No")

# 並非如此的普通狀況
else:
	IsAB = False # IsAB: 線段 AB 是否將點 C, D 分開？
	IsCD = False # IsCD: 線段 CD 是否將點 A, B 分開？
	if ans1 >= 0 and ans2 <= 0:
		IsAB = True
	if ans1 <= 0 and ans2 >= 0:
		IsAB = True
	if ans3 >= 0 and ans4 <= 0:
		IsCD = True
	if ans3 <= 0 and ans4 >= 0:
		IsCD = True
	
	# 輸出答案
	if IsAB == True and IsCD == True:
		print("Yes")
	else:
		print("No")

File: editorial_ch/chap4-1/prob4-1-5.c
#include <stdio.h>
#include <stdbool.h>

void swap(long long *x, long long *y) {
    long long tmp = *x;
    *x = *y;
    *y = tmp;
}

long long max(long long p1, long long p2) {
	if (p1 < p2) return p2;
	return p1;
}

long long min(long long p1, long long p2) {
	if (p1 < p2) return p1;
	return p2;
}

long long cross(long long ax, long long ay, long long bx, long long by) {
	// 向量 (ax, ay) 與 (bx, by) 的外積大小
	return ax * by - ay * bx;
}

int main() {
	// 輸入
	long long X1, Y1, X2, Y2, X3, Y3, X4, Y4;
	scanf("%lld%lld", &X1, &Y1); // 輸入點 A 的座標
	scanf("%lld%lld", &X2, &Y2); // 輸入點 B 的座標
	scanf("%lld%lld", &X3, &Y3); // 輸入點 C 的座標
	scanf("%lld%lld", &X4, &Y4); // 輸入點 D 的座標
	
	// 計算 cross(AB, AC)
	long long ans1 = cross(X2-X1, Y2-Y1, X3-X1, Y3-Y1);
	long long ans2 = cross(X2-X1, Y2-Y1, X4-X1, Y4-Y1);
	long long ans3 = cross(X4-X3, Y4-Y3, X1-X3, Y1-Y3);
	long long ans4 = cross(X4-X3, Y4-Y3, X2-X3, Y2-Y3);
	
	// 全部排在一直線上時（邊角案例）
	// 因為無法使用　pair 型態，與　C++ 的實作方法會稍有差異
	if (ans1 == 0 && ans2 == 0 && ans3 == 0 && ans4 == 0) {
		if (X1 == X2 && X2 == X3 && X3 == X4) {
			long long a = Y1, b = Y2; if (a > b) swap(&a, &b);
			long long c = Y3, d = Y4; if (c > d) swap(&c, &d);
			if (max(a, c) <= min(b, d)) printf("Yes\n");
			else printf("No\n");
		}
		else {
			long long a = X1, b = X2; if (a > b) swap(&a, &b);
			long long c = X3, d = X4; if (c > d) swap(&c, &d);
			if (max(a, c) <= min(b, d)) printf("Yes\n");
			else printf("No\n");
		}
		return 0;
	}
	
	// 並非如此時
	// IsAB: 線段 AB 是否將點 C, D 分開？
	// IsCD: 線段 CD 是否將點 A, B 分開？
	bool IsAB = false, IsCD = false;
	if (ans1 >= 0 && ans2 <= 0) IsAB = true;
	if (ans1 <= 0 && ans2 >= 0) IsAB = true;
	if (ans3 >= 0 && ans4 <= 0) IsCD = true;
	if (ans3 <= 0 && ans4 >= 0) IsCD = true;
	
	// 輸出答案
	if (IsAB == true && IsCD == true) {
		printf("Yes\n");
	}
	else {
		printf("No\n");
	}
	return 0;
}